### ISSUE TITLE
Wrong path when falling back to load-balancer

### ISSUE THREAD

**Issue by apeters1971** [xrootd/xrootd#2394]
I am redirecting a client asking for /a/b/c to a local path /x/y/z. 

The local path does not exist, so the client falls back to the initial endpoint and sets the correct tried information:
```

[2024-12-20 11:06:30.866676 +0100][Debug  ][XRootD            ] 	0. Redirected from: root://host.cern.ch:1094//a/b/c to: file://localhost/x/y/z
[2024-12-20 11:06:30.866676 +0100][Debug  ][XRootD            ] 	1. Retrying: root://host.cern.ch:1094//a/b/c
```

But what happens is that the clients sends the request with the wrong path for retry:

```
func=open                     path=/x/y/z info=tried=localhost&triedrc=srverr`
```

This is probably what should happen here!

**Comment by amadio** [comment]
I looked at this, but could not see where excatly the problem happens just from the messages above, as they come from a place separated from the code path, here:

https://github.com/xrootd/xrootd/blob/71ea4b5b2c68937500f4c7b70a9d775781776eb5/src/XrdCl/XrdClXRootDMsgHandler.hh#L91-L106

I suppose the code path where the error happens is from within `HandleError` in the statement below (due to `EntryRedirect` type):

https://github.com/xrootd/xrootd/blob/71ea4b5b2c68937500f4c7b70a9d775781776eb5/src/XrdCl/XrdClXRootDMsgHandler.cc#L752

at the end of the processing block for `kXR_redirect`, which eventually reaches this block:

https://github.com/xrootd/xrootd/blob/71ea4b5b2c68937500f4c7b70a9d775781776eb5/src/XrdCl/XrdClXRootDMsgHandler.cc#L2126-L2149

where we probably need to make a change. However, there are other similar code paths, so to be able to reproduce the same code path as described above we need more information, like a full Debug or Dump level client log. Otherwise, it's hard to make progress on this issue. Once I understand better how the local redirect falls back to the load balancer I can create a test and fix it. However, redirects to `file://localhost/` sound like something that we should actually restrict to improve security, so I'm not sure we should actually fix this instead of adding the restriction. There's already a check trying to handle this in part here:

https://github.com/xrootd/xrootd/blob/71ea4b5b2c68937500f4c7b70a9d775781776eb5/src/XrdCl/XrdClXRootDMsgHandler.cc#L715-L728

So it seems that maybe we should remove that `!newUrl.IsLocalFile()` from the condition instead and treat it as an error.

**Comment by smithdh** [comment]
At first leaving asside the question of whether redirecting to file:// is good or not, or should even be disallowed: This seems to be a particular case where a redirect specifies a new path name, not only a new host or extra cgi.  If that redirect fails and the client determines it can retry at a load-balancer (manager) it should probably return to using the original path name, whereas I believe it is still sending the new pathname. Am I right to think the correct behavior is to use the original pathname used at that load-balancer?

**Comment by amadio** [comment]
@smithdh I believe you are correct, the client should revert to the original path on retry. Do you plan to submit the commit above as a fix for this issue? Please let me know.

**Comment by dynamic-entropy** [comment]
> redirecting to file:// is good or not, or should even be disallowed

@smithdh 
Isn't it the reason `XrdCmsRedirLocal` was written? To support redirecting to locally mounted filesystems? https://github.com/xrootd/xrootd/blob/e5e9c284138bed15860474e4b1794bac29d0c26a/src/XrdCms/XrdCmsRedirLocal.hh#L21-L27

**Comment by amadio** [comment]
@dynamic-entropy Yes, you are right, however, we can redirect to `root://localhost/...` for that instead of `file://...`. So it's something we should think about as security measure in any case. At least with `root://` the authentication can be enforced.

**Comment by dynamic-entropy** [comment]
> we can redirect to root://localhost/... for that instead of file://....

No, this is not TPC. There is no server listening on the worker node ( localhost ). We do not need to open another xroot session, since if something were listening on localhost, it would be the same redirector, resulting in a loop.

**Comment by amadio** [comment]
What I'm saying is that the local node should run an XRootD server to be able to receive such redirects. Anyway, let's discuss this offline, it's offtopic in this issue.
